src/Edit.c

```diff
@@ -2199,6 +2199,173 @@ void EditShowHex(void) {
+		NotifyRectangleSelection();
+		return;
+	}
+	EncodingTable[62] = (UrlSafeEncode) ? '-' : '+';
+	EncodingTable[63] = (UrlSafeEncode) ? '_' : '/';
+	size_t len = SciCall_GetSelTextLength();
+	size_t olen = 1 + ((len + ( (len % 3 > 0) ? 3 - (len % 3) : 0 ) / 3 ) * 4);
+	char *in = (char *)NP2HeapAlloc(len+1);
+	SciCall_GetSelText(in);
+	char *out = (char *)NP2HeapAlloc(olen);
+	BinToBase64(out, in, len);
+	EditReplaceMainSelection(strlen(out),out);
+	NP2HeapFree(in);
+	NP2HeapFree(out);
+}
+
+//=============================================================================
+//
+// EditBase64Decode()
+//
+
+//Added '-_' characters in DecodingTable to correctly interpret the 'URL Safe Encode'
+static char const DecodingTable[] = {
+	64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+	64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+	64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 62, 64, 62, 64, 63,
+	52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 64, 64, 64, 65, 64, 64,
+	64,  0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14,
+	15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 64, 64, 64, 64, 63,
+	64, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40,
+	41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 64, 64, 64, 64, 64,
+	64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+	64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+	64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+	64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+	64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+	64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+	64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+	64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64
+};
+
+static void* Base64ToBin( void* dest, char const* src) {
+	unsigned char const* s = (unsigned char*)src;
+	char* p = dest;
+	for(;;) {
+		int const a = DecodingTable[ *s ];
+		if ( a == 64 || a == 65 ) return p;
+		int const b = DecodingTable[ *++s ];
+		if ( b == 64 || b == 65 ) return 0;
+		*p++ = ( a << 2u ) | ( b >> 4u );
+		int const c = DecodingTable[ *++s ];
+		if ( c == 64 ) return 0;
+		int const d = DecodingTable[ *++s ];
+		if ( d == 64 ) return 0;
+		if ( c == 65 ) {
+			if ( d != 65 ) return 0;
+				return p;
+		}
+		*p++ = ( b << 4u ) | ( c >> 2u );
+		if ( d == 65 ) return p;
+		*p++ = ( c << 6u ) | ( d >> 0u );
+		++s;
+	}
+	return p;
+}
+
+void EditBase64Decode(bool DecodeAsHex){
+	if (SciCall_IsRectangleSelection()){
+		NotifyRectangleSelection();
+		return;
+	}
+	size_t len = SciCall_GetSelTextLength();
+	char *in = (char *)NP2HeapAlloc(len+1);
+	SciCall_GetSelText(in);
+	size_t olen = len / 4 * 3;
+	for (int i=len; i-->0; ) {
+		if (in[i] == '=') {
+			olen--;
+		} else {
+			break;
+		}
+	}
+	char *out = (char *)NP2HeapAlloc(olen);
+	Base64ToBin(out, in);
+	if(DecodeAsHex){
+		char *cch = (char *)NP2HeapAlloc((len * 3) + 4 + (floor(len / 16) * 1));

```

Comment:
clang fails on `floor`, see https://github.com/zufuliu/notepad2/actions/runs/3601306960/jobs/6066952554#step:5:210
how about `len*3  + 2*(len/16 + 1) + 1` => `len*3 + len/8 + 3` ?

Reply:
I would say it should be fine. Shall I make the change or are you going to make it?

Reply:
multiply by 2 is due to CRLF line endings, you can push a change (without handing different eol) to make CI happy.

The current code doesn't looks good (e.g. `Blkx` can be inlined?), but it hard to write optimized base64 encoding/decoding code.

Reply:
Fixed what you requested 😉

Reply:
I think inline not means make them one line, but remove them completely?

Reply:
I have tried other algorithms but this seems to be the only one that really works... I have no alternative ideas at the moment...

Reply:
OK, I can rework the code in this weekend.